### PR TITLE
Testing improvements and minor fixes.

### PR DESCRIPTION
**Pull request by hameerabbasi**
## What type of PR is this? (check all applicable)

- [x] 💾 Refactor
- [ ] 🪄 Feature
- [ ] 🐞 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📚 Documentation
- [x] 🧪 Test
- [ ] 🛠️ Other

## Checklist

- [x] Code follows style guide
- [x] Tests added
- [ ] Documented the changes

### CODE REVIEW

File: sparse/mlir_backend/tests/test_simple.py [hunk 1]
```diff
@@ -108,93 +123,92 @@ def test_dense_format(dtype, shape):
...
+
+    assert_array_equal(sp_arr, sp_arr_roundtripped)
 
 
 @parametrize_dtypes
-def test_add(rng, dtype):
+@parametrize_scipy_fmt
+def test_add(rng, dtype, format):
+    if format == "coo":
+        pytest.xfail(reason="https://github.com/llvm/llvm-project/issues/116012")
```
Comment (mtsokol) [inline sparse/mlir_backend/tests/test_simple.py:193]: This skip is about converting COO, here we're testing simple addition - we should test COO+COO.

We should also add dense to parametrization and test addition for different formats as before (now test is narrower than before): CSC+CSR, Dense+CSR, etc.

File: sparse/mlir_backend/tests/test_simple.py [hunk 2]
```diff
@@ -108,93 +123,92 @@ def test_dense_format(dtype, shape):
...
+    assert_array_equal(sp_arr, sp_arr_roundtripped)
 
-    dense_returned = sparse.to_numpy(dense_tensor)
-    np.testing.assert_equal(dense_returned, dense)
 
-    coo_returned = sparse.to_scipy(coo_tensor)
-    np.testing.assert_equal(coo_returned.todense(), coo.todense())
+@parametrize_dtypes
+def test_roundtrip_dense(rng, dtype):
+    SHAPE = (80, 100)
```
Comment (mtsokol) [inline sparse/mlir_backend/tests/test_simple.py:175]: For dense let's also parametrize shape, and consider 1D and 3D.

**Review by mtsokol — CHANGES_REQUESTED** [review]

File: sparse/mlir_backend/tests/test_simple.py (hunk 1 above)
Comment (hameerabbasi) [reply sparse/mlir_backend/tests/test_simple.py:193]: We need to use `.asformat` under the current framework to get to COO output. Therefore, this fails for the same reason.

File: sparse/mlir_backend/tests/test_simple.py (hunk 2 above)
Comment (hameerabbasi) [reply sparse/mlir_backend/tests/test_simple.py]: Done in https://github.com/pydata/sparse/pull/808/commits/406330b082d9a926095874e588c5db630394acaf.

File: sparse/mlir_backend/tests/test_simple.py (hunk 1 above)
Comment (mtsokol) [reply sparse/mlir_backend/tests/test_simple.py:193]: So this means that we won't test COO at all?

Comment (hameerabbasi) [reply sparse/mlir_backend/tests/test_simple.py:193]: Unfortunately we can't; conversion is a fundamental operation and it's broken. 😞

Comment (mtsokol) [reply sparse/mlir_backend/tests/test_simple.py:193]: So the output of COO+COO is now DCSR. How about converting it to dense for now and performing assertion? Then we can test this operation for COO.

Comment (hameerabbasi) [reply sparse/mlir_backend/tests/test_simple.py:193]: I just tried testing this op by converting to dense:

<details><summary>Code</summary>

```python
    # TODO: Remove condition after https://github.com/llvm/llvm-project/issues/116012 is fixed
    if format != "coo":
        actual_sps = sparse.to_scipy(actual.asformat(sp_arr1.format))
        assert_sps_equal(expected, actual_sps, check_canonical=True)
    else:
        out_fmt = sparse.levels.get_storage_format(
            levels=(sparse.levels.Level(sparse.levels.LevelFormat.Dense),) * 2,
            order="C",
            pos_width=64,
            crd_width=64,
            dtype=dtype,
        )
        expected = expected.todense()
        actual = sparse.to_numpy(actual.asformat(out_fmt))
        np.testing.assert_array_equal(expected, actual)
```
</details> 

But I still get failures:

<details><summary>Example</summary>

```log
______________________________ test_add[coo-int8] ______________________________

rng = Generator(PCG64) at 0x1401A11C0, dtype = <class 'numpy.int8'>
format = 'coo'

    @parametrize_dtypes
    @parametrize_scipy_fmt
    def test_add(rng, dtype, format):
        SHAPE = (100, 50)
        DENSITY = 0.5
        sampler = generate_sampler(dtype, rng)
        sps_arr1 = sps.random_array(
            SHAPE, density=DENSITY, format=format, dtype=dtype, random_state=rng, data_sampler=sampler
        )
        sps_arr2 = sps.random_array(
            SHAPE, density=DENSITY, format=format, dtype=dtype, random_state=rng, data_sampler=sampler
        )
    
        sp_arr1 = sparse.asarray(sps_arr1)
        sp_arr2 = sparse.asarray(sps_arr2)
    
        expected = sps_arr1 + sps_arr2
        actual = sparse.add(sp_arr1, sp_arr2)
    
        # TODO: Remove condition after https://github.com/llvm/llvm-project/issues/116012 is fixed
        if format != "coo":
            actual_sps = sparse.to_scipy(actual.asformat(sp_arr1.format))
            assert_sps_equal(expected, actual_sps, check_canonical=True)
        else:
            out_fmt = sparse.levels.get_storage_format(
                levels=(sparse.levels.Level(sparse.levels.LevelFormat.Dense),) * 2,
                order="C",
                pos_width=64,
                crd_width=64,
                dtype=dtype,
            )
            expected = expected.todense()
            actual = sparse.to_numpy(actual.asformat(out_fmt))
>           np.testing.assert_array_equal(expected, actual)

sparse/mlir_backend/tests/test_simple.py:222: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.pixi/envs/mlir-dev/lib/python3.13/site-packages/numpy/_utils/__init__.py:85: in wrapper
    return fun(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<built-in function eq>, array([[ -6,  -1, -10, ...,  -5,   0,   0],
       [  0,  -6,  -6, ...,   3,   0,   0],
     ...,  -1,  -6,   0],
       [  4,   0,   0, ...,   4,  -3,   0],
       [ -6,   0,   9, ...,   0,   0,  -5]], dtype=int8))
kwds = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError: 
E           Arrays are not equal
E           
E           Mismatched elements: 3993 / 5000 (79.9%)
E           Max absolute difference among violations: 30
E           Max relative difference among violations: 19.
E            ACTUAL: array([[ -6,  -1, -10, ...,  -5,   0,   0],
E                  [  0,  -6,  -6, ...,   3,   0,   0],
E                  [  8,  -7,  -3, ...,   0,   0,  -2],...
E            DESIRED: array([[-10,   0,  10, ...,  -1,   0,   0],
E                  [  0,  -9,   9, ...,  -2,   9,   0],
E                  [  7,   9,  -6, ...,  -8,   7,   0],...

.pixi/envs/mlir-dev/lib/python3.13/contextlib.py:85: AssertionError
```
</details>

Comment (mtsokol) [reply sparse/mlir_backend/tests/test_simple.py:193]: But the conversion to dense happens on DCSR, right?

Comment (hameerabbasi) [reply sparse/mlir_backend/tests/test_simple.py:193]: Yes, although I suspect `add` is broken on COO.

Comment (mtsokol) [reply sparse/mlir_backend/tests/test_simple.py:193]: Uh, Ok, let's troubleshoot it separately. In the previous `test_add` version `COO+COO=COO` was present and tested so we know it works. We need to check `COO+COO=DCSR` and convertion `DCSR=>Dense` separately.

File: sparse/mlir_backend/tests/test_simple.py [hunk 3]
```diff
@@ -108,93 +123,109 @@ def test_dense_format(dtype, shape):
...
+        pytest.xfail(reason="https://github.com/llvm/llvm-project/issues/116012")
     SHAPE = (100, 50)
     DENSITY = 0.5
     sampler = generate_sampler(dtype, rng)
+    sps_arr1 = sps.random_array(
+        SHAPE, density=DENSITY, format=format, dtype=dtype, random_state=rng, data_sampler=sampler
+    )
+    sps_arr2 = sps.random_array(
+        SHAPE, density=DENSITY, format=format, dtype=dtype, random_state=rng, data_sampler=sampler
+    )
```
Comment (mtsokol) [inline sparse/mlir_backend/tests/test_simple.py:202]: Can we still test adding different formats? `CSR+CSC`, `CSR+Dense`? We had it previously in the test.

**Review by mtsokol — CHANGES_REQUESTED** [review]

File: sparse/mlir_backend/tests/test_simple.py (hunk 3 above)
Comment (hameerabbasi) [reply sparse/mlir_backend/tests/test_simple.py]: Done in https://github.com/pydata/sparse/pull/808/commits/0b910b93f6e12ba74bc8d7bdff7b7b5f6d220794. As a side-note, almost every case with COO segfaulted with `add`; so I really do think it is `add` failing on COO and not `asformat`.

**Review by mtsokol — APPROVED** [review]

**Pull request merged**